magicgui/_magicgui.py

```diff
@@ -1,18 +1,70 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Callable, Optional, Union, overload
+import inspect
+from functools import partial
+from types import FunctionType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
 from warnings import warn
 
 from typing_extensions import Literal
 
+from magicgui.widgets import FunctionGui, MainFunctionGui
+
 if TYPE_CHECKING:
     from magicgui.application import AppRef
-    from magicgui.widgets import FunctionGui, MainFunctionGui
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+
+
+def _magicgui(function=None, factory=False, main_window=False, **kwargs):
+    """Actual private magicui decorator.
+
+    if factory is `True` will return a MagicFactory instance, that can be called
+    to return a `FunctionGui` instance.  See docstring of ``magicgui`` for parameters
+    """
+    if "result" in kwargs["param_options"]:
+        warn(
+            "\n\nThe 'result' option is deprecated and will be removed in the future."
+            "Please use `result_widget=True` instead.\n",
+            FutureWarning,
+        )
+
+        kwargs["param_options"].pop("result")
+        kwargs["result_widget"] = True
+
+    def inner_func(func: Callable) -> Union[FunctionGui, MagicFactory]:
+        if not callable(func):
+            raise TypeError("the first argument must be callable")
+
+        magic_class = MainFunctionGui if main_window else FunctionGui
+
+        if factory:
+            return MagicFactory(func, magic_class=magic_class, **kwargs)
+        return magic_class(func, **kwargs)
```

Comment:
This logic I find confusing... Why isn't this (forgive the names):

```python
a_factory = MagicFactory(func, magic_class=magic_class, **kwargs)
if factory:
    return a_factory
else:
    return a_factory()
```

Reply:
I guess I felt like there was no need to go through the whole `MagicFactory.__new__` instantiation and stuff if we knew we were immediately going to just the instance.

Reply:
Right! Maybe add a comment like, "magic factory needs to be a lot more magical than necessary if we are instantiating straight away, so we shortcut that here"?

Reply:
done